pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java

```diff
@@ -100,6 +102,8 @@ public ServerCnx(BrokerService service) {
         this.producers = new ConcurrentLongHashMap<>(8, 1);
         this.consumers = new ConcurrentLongHashMap<>(8, 1);
         this.replicatorPrefix = service.pulsar().getConfiguration().getReplicatorPrefix();
+        this.nonPersistentMessageSemaphore = new Semaphore(
```

Comment:
so, you mean instead semaphore we can have just a counter like `pendingSendRequest`, but then we need to decrement this counter on `completedSendOperation()` and that called from a different thread. 
so, it might not be a thread-safe and it might not throttle with accurate configured number? right?